File: smalivm/src/test/java/org/cf/smalivm/configuration/ConfigurationTest.java
package org.cf.smalivm.configuration;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.cf.smalivm.dex.SmaliClassLoader;
import org.cf.util.ClassNameUtils;

import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.jar.JarEntry;
import java.util.jar.JarInputStream;
import java.util.stream.Collectors;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class ConfigurationTest {

    private Configuration configuration;

    @BeforeEach
    public void setUp() {
        configuration = Configuration.instance();
    }

    @Test
    public void stringIsImmutable() {
        assertTrue(configuration.isImmutable("Ljava/lang/String;"));
    }

    @Test
    public void stringArrayIsImmutable() {
        assertFalse(configuration.isImmutable("[Ljava/lang/String;"));
    }

    @Test
    public void stringBuilderIsImmutable() {
        assertFalse(configuration.isImmutable("Ljava/lang/StringBuilder;"));
    }

    @Test
    public void safeClassesCanBeLoaded() throws IOException, ClassNotFoundException {
        ensureClassesExist(Configuration.SAFE_CLASSES_PATH);
    }

    @Test
    public void immutableClassesCanBeLoaded() throws IOException, ClassNotFoundException {
        List<String> lines = Configuration.load(Configuration.IMMUTABLE_CLASSES_PATH);
        // Ignore primitives and inner classes
        List<String> classNames = lines.stream().filter(l -> l.startsWith("L") && !l.contains("$")).collect(Collectors.toList());
        ensureClassesExist(classNames, Configuration.IMMUTABLE_CLASSES_PATH, SmaliClassLoader.FRAMEWORK_STUBS_JAR);
    }

    @Test
    public void safeFrameworkClassesExist() throws IOException, ClassNotFoundException {
        List<String> lines = Configuration.load(Configuration.SAFE_FRAMEWORK_CLASSES_PATH);
        ensureClassesExist(lines, Configuration.SAFE_FRAMEWORK_CLASSES_PATH, SmaliClassLoader.FRAMEWORK_STUBS_JAR);
    }

    @Test
    public void frameworkClassesExist() throws IOException, ClassNotFoundException {
        List<String> lines = Configuration.load(Configuration.FRAMEWORK_CLASSES_PATH);
        List<String> configClasses = lines.stream().map(l -> l.substring(0, l.indexOf(':'))).collect(Collectors.toList());
        ensureClassesExist(configClasses, Configuration.FRAMEWORK_CLASSES_PATH, SmaliClassLoader.FRAMEWORK_STUBS_JAR);
    }

    private void ensureClassesExist(String configResPath) throws ClassNotFoundException {
        List<String> classNames = Configuration.load(configResPath);
        for (String className : classNames) {
            String sourceName = ClassNameUtils.internalToSource(className);
            ConfigurationTest.class.getClassLoader().loadClass(sourceName);
        }
    }

    private void ensureClassesExist(List<String> classNames, String configResPath, String jarResourcePath) throws IOException {
        InputStream is = ConfigurationTest.class.getResourceAsStream(jarResourcePath);
        JarInputStream jis = new JarInputStream(is);
        JarEntry je;

        Set<String> jarClasses = new HashSet<String>();
        while ((je = jis.getNextJarEntry()) != null) {
            String entryName = je.getName();
            if (entryName.endsWith(".class")) {
                String className = "L" + entryName.substring(0, entryName.indexOf('.')) + ";";
                jarClasses.add(className);
            }
        }

        List<String> nonExistentClasses = classNames.stream().filter(l -> !jarClasses.contains(l)).collect(Collectors.toList());

        String msg = "Classes in " + configResPath + " don't exist: " + Arrays.toString(nonExistentClasses.toArray());
        assertEquals(0, nonExistentClasses.size(), msg);
    }
}

File: smalivm/src/test/java/org/cf/smalivm/configuration/ConfigurationTest.kt
package org.cf.smalivm.configuration

import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.util.ClassNameUtils
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.IOException
import java.util.*
import java.util.jar.JarEntry
import java.util.jar.JarInputStream
import java.util.stream.Collectors

class ConfigurationTest {
    private var configuration: Configuration? = null
    @BeforeEach
    fun setUp() {
        configuration = Configuration.instance()
    }

    @Test
    fun stringIsImmutable() {
        Assertions.assertTrue(configuration!!.isImmutable("Ljava/lang/String;"))
    }

    @Test
    fun stringArrayIsImmutable() {
        Assertions.assertFalse(configuration!!.isImmutable("[Ljava/lang/String;"))
    }

    @Test
    fun stringBuilderIsImmutable() {
        Assertions.assertFalse(configuration!!.isImmutable("Ljava/lang/StringBuilder;"))
    }

    @Test
    @Throws(IOException::class, ClassNotFoundException::class)
    fun safeClassesCanBeLoaded() {
        ensureClassesExist(Configuration.SAFE_CLASSES_PATH)
    }

    @Test
    @Throws(IOException::class)
    fun immutableClassesCanBeLoaded() {
        val lines = Configuration.load(Configuration.IMMUTABLE_CLASSES_PATH)
        // Ignore primitives and inner classes
        val classNames = lines.stream().filter { l: String -> l.startsWith("L") && !l.contains("$") }.collect(Collectors.toList())
        ensureClassesExist(classNames, Configuration.IMMUTABLE_CLASSES_PATH, SmaliClassLoader.FRAMEWORK_STUBS_JAR)
    }

    @Test
    @Throws(IOException::class)
    fun safeFrameworkClassesExist() {
        val lines = Configuration.load(Configuration.SAFE_FRAMEWORK_CLASSES_PATH)
        ensureClassesExist(lines, Configuration.SAFE_FRAMEWORK_CLASSES_PATH, SmaliClassLoader.FRAMEWORK_STUBS_JAR)
    }

    @Test
    @Throws(IOException::class)
    fun frameworkClassesExist() {
        val lines = Configuration.load(Configuration.FRAMEWORK_CLASSES_PATH)
        val configClasses = lines.stream().map { l: String -> l.substring(0, l.indexOf(':')) }.collect(Collectors.toList())
        ensureClassesExist(configClasses, Configuration.FRAMEWORK_CLASSES_PATH, SmaliClassLoader.FRAMEWORK_STUBS_JAR)
    }

    @Throws(ClassNotFoundException::class)
    private fun ensureClassesExist(configResPath: String) {
        val classNames = Configuration.load(configResPath)
        for (className in classNames) {
            val sourceName = ClassNameUtils.internalToSource(className)
            ConfigurationTest::class.java.classLoader.loadClass(sourceName)
        }
    }

    @Throws(IOException::class)
    private fun ensureClassesExist(classNames: List<String>, configResPath: String, jarResourcePath: String) {
        val inputStream = ConfigurationTest::class.java.getResourceAsStream(jarResourcePath)
        val jis = JarInputStream(inputStream)
        var je: JarEntry?
        val jarClasses: MutableSet<String> = HashSet()
        while (true) {
            je = jis.nextJarEntry
            if (je == null) {
                break
            }
            val entryName = je.name
            if (entryName.endsWith(".class")) {
                val className = "L" + entryName.substring(0, entryName.indexOf('.')) + ";"
                jarClasses.add(className)
            }
        }
        val nonExistentClasses = classNames.stream().filter { l: String -> !jarClasses.contains(l) }.collect(Collectors.toList())
        val msg = "Classes in $configResPath don't exist: ${nonExistentClasses.toTypedArray().contentToString()}"
        Assertions.assertEquals(0, nonExistentClasses.size, msg)
    }
}
